### PR TITLE
ITF trace reader - `Either`-based alternative

### LINKED ISSUE #2272: Implement a reader for ITF -> TlaIR
We need to be able to read ITF files as a sequence of `State`s in the IR (e.g. for trace evaluation).

### PR DESCRIPTION
**Pull request by Kukovec**
- [x] Tests added for any new code
- [x] Ran `make fmt-fix` (or had formatting run automatically on all files edited)
- [ ] Documentation added for any new functionality
- [ ] [Entries added to `./unreleased/`][changelog format] for any new functionality

closes #2272

Alternative to #2273, uses `Either`.

[changelog format]: https://github.com/informalsystems/apalache/blob/main/CONTRIBUTING.md#how-to-record-a-change

### CODE REVIEW

File: tla-io/src/test/scala/at/forsyte/apalache/io/itf/TestItfToTla.scala [hunk 1]
```diff
@@ -0,0 +1,480 @@
...
+    }
+
+    val nat = singleUS("Nat")
+
+    assert {
+      itfToTla.attemptUnserializable(nat).exists(_.isLeft)
+    }
+  }
+
+  test("typeDrivenBuild") {
```
Comment (shonfeder) [inline tla-io/src/test/scala/at/forsyte/apalache/io/itf/TestItfToTla.scala:133]: This test case is really long. We get much better feedback from tests if we can break them into smaller units, so that when we get "test case 'such and such property should hold' failed" we know from the name of the test what has failed.

It's not essential, but imo these tests will be much more helpful in case of breakage if they are broken into units that note what they are testing. Seems like the natural categorization based on your current tests is to break each type into its own test case?

This basically just amounts to replacing the (tbh, slightly cryptic ;) ) comments with test descriptions.

File: tla-io/src/test/scala/at/forsyte/apalache/io/itf/TestItfToTla.scala [hunk 2]
```diff
@@ -0,0 +1,480 @@
...
+          .map(_.build)
+          .contains(tla
+                .setAsFun(tla.enumSet(
+                        Seq(1, 2).map { i => tla.tuple(tla.int(i), tla.int(i)) }: _*
+                    ))
+                .build))
+
+  }
+
+  test("getTrace") {
```
Comment (shonfeder) [inline tla-io/src/test/scala/at/forsyte/apalache/io/itf/TestItfToTla.scala:361]: I don't see any tests here to check that the correct behavior occurs in the presence of junk fields, like `{"#set": [], "foo": "bar"}`.

Can you add a named test case for this (and similar edge cases, if you think of any?).

File: tla-io/src/main/scala/at/forsyte/apalache/io/itf/ItfToTla.scala [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+                      )
+                      bi <- typeDrivenBuild(json.getFieldOpt(k).get, fieldTypes(k))
+                    } yield k -> bi
+                  }
+              )
+            // Records must contain at least 1 kv pair
+            _ <- requirement(
+                recElems.nonEmpty,
+                ItfFormatError(s"Records may not be empty."),
+            )
```
Comment (shonfeder) [inline tla-io/src/main/scala/at/forsyte/apalache/io/itf/ItfToTla.scala:192]: This `requirement` idiom reads really nice to me!

**Review by shonfeder — APPROVED** [review]
Code LGTM!

One thing I'd still like to see here before merging: put the ITF field constants in place where they can be shared with the writer code. And we can replace the magic strings in the writer with reference to the constants in a followup.

I also find the test cases to be much larger than is preferable, and left some notes indicating what I think would help. But if you disagree or don't think it's worth the effort, I'm fine with merging as is.

Thanks for the discussions!

**Comment by Kukovec** [comment]
> One thing I'd still like to see here before merging: put the ITF field constants in place where they can be shared with the writer code. And we can replace the magic strings in the writer with reference to the constants in a followup.

Yes, I have that planned as a followup. I'll open an issue for that.

File: tla-io/src/main/scala/at/forsyte/apalache/io/itf/ItfToTla.scala [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+  val VAR_TYPES_FIELD: String = "varTypes"
+  val VARS_FIELD: String = "vars"
+  val STATES_FIELD: String = "states"
+
+  val BIG_INT_FIELD: String = "#bigint"
+  val TUP_FIELD: String = "#tup"
+  val SET_FIELD: String = "#set"
+  val MAP_FIELD: String = "#map"
+  val UNSERIALIZABLE_FIELD: String = "#unserializable"
+}
```
[the remark this answers is not in the document]
Comment (Kukovec) [reply tla-io/src/main/scala/at/forsyte/apalache/io/itf/ItfToTla.scala:315]: I'll move it and import it in the writer, see #2284.

File: tla-io/src/test/scala/at/forsyte/apalache/io/itf/TestItfToTla.scala (hunk 1 above)
Comment (Kukovec) [reply tla-io/src/test/scala/at/forsyte/apalache/io/itf/TestItfToTla.scala:133]: Sure, I can make tighter groups, np.
